### ISSUE TITLE
NodePort creation

### ISSUE THREAD

**Issue by kwill4026** [mongodb/mongodb-enterprise-kubernetes#14]
In regards to the mongodb-enterprise-kubernetes, I wanted to know how you all were exposing the second service (my-replica-set-svc-external) as a NodePort for external connections to access the pods. If you could, please explain how you are achieving this in the code. Is a second service being created via a .yaml file in order to declare a NodePort?

**Comment by antonlisovenko** [comment]
The Operator creates two services each time - the first one for internal connectivity - the second one exposing the port on the node where the pod runs. The operator doesn't use any yaml files, just the K8s API.
Note that the connectivity to Mongo databases from outside of K8s cluster is a problem right now as needs support from the MongoDB server, the changes are being developed currently

**Comment by kwill4026** [comment]
Thanks for the response Anton. Is this a problem only from the mongodb-kubernetes operator or a problem that also exists in other MongoDB replicaset projects all around attempting to expose the port for external comms?

**Comment by antonlisovenko** [comment]
This is a general problem which will affect any MongoDB running on Kubernetes. The communication between MongoDB drivers and database members requires that the databases know their hostnames and return them to the drivers which can not work if the node is addressed using a hostname which isn’t the same as those used by other members of the replica set.
We are working on support for external connectivity, and plan to support it for v1.0 of the Operator.

**Comment by kwill4026** [comment]
Nice. I only ask because aside from the mnogodb-kubernetes-operator, I was running my own mongodb statefulset and was wondering how I could expose it to external connections. Therefore I created a second service via my yaml and gave the type as a NodePort which enables external connections but does not always connect to the master upon the first try. It acts more as a load balancer, picking which node to send to with each connection request, so I dont always get the same node. Sometimes I get the replicaset instead and have to exit and try the connection again until it lands on the primary.

**Comment by abinasroy** [comment]
how when mongo ent opertor will have GA release.will it support TLS?

**Comment by antonlisovenko** [comment]
> how when mongo ent opertor will have GA release.will it support TLS?
Yes it will. We aim the GA for the middle of the year.
For future - feel free to ask any questions in our [slack channel](https://mongo-db.slack.com/messages/CB323LCG5/)

**Comment by jamesbroadhead** [comment]
I think that this question has been answered, and we plan future work to support external access to Replica Sets in the future. 
As a result, I'm closing this issue.

**Comment by Symbianx** [comment]
@antonlisovenko 
Is this supported on 1.0? We deployed a cluster and only a ClusterIP service was created.

**Comment by antonlisovenko** [comment]
@Symbianx you need to specify `exposeExternally` field:
https://docs.mongodb.com/kubernetes-operator/stable/reference/k8s-operator-specification/#standalone-settings

**Comment by Symbianx** [comment]
@antonlisovenko 
Sorry, I was referring to replica-set deployments, which does not appear to have that setting. Is there a way to do it?

**Comment by antonlisovenko** [comment]
@Symbianx it does have it and the NodePort service will be created - it will just not give any benefit as of now as replica sets cannot be connected from outside using NodePort - only standalones and sharded clusters can

**Comment by kwill4026** [comment]
@antonlisovenko so the inability to connect from outside using NodePort is only constrained to replica-sets currently and you are saying we should be able to use a product such as MongoDB Compass to automatically connect from outside using a NodePort if its a standalone or sharded?

I definitely have tried setting up a replicaset via the Mongodb enterprise Kubernetes Operator, exposed the service as a node port and was unable to connect via compass with the full replicaset URI. However when I specified only the primary and no secondaries I was able to connect directly with no issues. 

The maintainers of this chart (https://github.com/helm/charts/tree/master/stable/mongodb) claim to figure this issue out by exposing MongoDB externally using the NGINX Ingress Controller. I have not had time to set this up so I am not sure if this will solve the kubernetes and mongodb replicaset external connection issues.
